File: src/lexer.rs
use std::cell::RefCell;
use std::ops::Add;
use crate::error::error;
use crate::token::{Token, TokenType, make_keywords};

pub fn scan(input: String, tokens: &mut RefCell<Vec<Token>>, repl: bool) {
    let symbols = "\\|/?><!#@`^~%&*-_+=";
    let keywords = make_keywords();
    let mut input: RefCell<String> = RefCell::new(input.chars().rev().collect());
    let mut x = 0;
    let mut y = 0;
    'outer: loop {
        let c = match input.get_mut().pop() {
            Some(c) => c,
            None => break
        };
        match c {
            '@' => tokens.get_mut().push(
                Token { token: TokenType::LAMBDA, content: "@".to_string(), span: (y, x) }),
            '^' => tokens.get_mut().push(
                Token { token: TokenType::RETURN, content: "^".to_string(), span: (y, x) }),
            '|' => tokens.get_mut().push(
                Token { token: TokenType::DELIMITER, content: "|".to_string(), span: (y, x) }),
            ',' => tokens.get_mut().push(
                Token { token: TokenType::COMMA, content: ",".to_string(), span: (y, x) }),
            '.' => tokens.get_mut().push(
                Token { token: TokenType::DOT, content: ".".to_string(), span: (y, x) }),
            ':' => {
                let c = match input.get_mut().pop() {
                    Some(c) => c,
                    None => break
                };
                if c == '=' {
                    tokens.get_mut().push(
                        Token { token: TokenType::BIND, content: ":=".to_string(), span: (y, x) });
                    x += 1;
                } else {
                    input = RefCell::new(input.take().add(c.to_string().as_str()));
                    tokens.get_mut().push(
                        Token { token: TokenType::ATOM, content: ":".to_string(), span: (y, x) })
                }
            }
            '=' => tokens.get_mut().push(
                Token { token: TokenType::EQ, content: "=".to_string(), span: (y, x) }),
            '(' | '[' => tokens.get_mut().push(
                Token { token: TokenType::LPAREN, content: "(".to_string(), span: (y, x) }),
            ')' | ']' => tokens.get_mut().push(
                Token { token: TokenType::RPAREN, content: ")".to_string(), span: (y, x) }),
            '"' => {
                let mut string: String = String::from(c);
                loop {
                    let c = match input.get_mut().pop() {
                        Some(c) => c,
                        None => break 'outer,
                    };
                    if c != '"' {
                        string = string.add(c.to_string().as_str());
                    } else {
                        input = RefCell::new(input.take().add(c.to_string().as_str()));
                        break;
                    }
                }
                let len = string.len() as i64;
                tokens.get_mut().push(
                    Token { token: TokenType::STRING, content: string, span: (y, x) });
                x += len;
                continue;
            }
            '\r' | '\t' | ' ' => {}
            '\n' => {
                x = 0;
                y += 1;
                tokens.get_mut().push(
                    Token { token: TokenType::COMMA, content: ",".to_string(), span: (y, x) });
            }
            _ if c.is_digit(10) => {
                let mut num: String = String::from(c);
                loop {
                    let c = match input.get_mut().pop() {
                        Some(c) => c,
                        None => break 'outer,
                    };
                    if c.is_digit(10) || c == '.' {
                        num = num.add(c.to_string().as_str());
                    } else {
                        input = RefCell::new(input.take().add(c.to_string().as_str()));
                        break;
                    }
                }
                let len = num.len() as i64;
                tokens.get_mut().push(
                    Token { token: TokenType::NUM, content: num, span: (y, x) });
                x += len;
                continue;
            }
            _ if c.is_alphabetic() || symbols.contains(c) => {
                let mut name: String = String::from(c);
                loop {
                    let c = match input.get_mut().pop() {
                        Some(c) => c,
                        None => break 'outer,
                    };
                    if c.is_alphabetic() || symbols.contains(c) {
                        name = name.add(c.to_string().as_str());
                    } else {
                        input = RefCell::new(input.take().add(c.to_string().as_str()));
                        break;
                    }
                }
                let len = name.len() as i64;
                if keywords.contains_key(&*name) {
                    let token = *keywords.get(&*name).unwrap();
                    tokens.get_mut().push(
                        Token { token, content: name, span: (y, x) });
                } else if name == "NB" { // comment
                    loop {
                        let c = match input.get_mut().pop() {
                            Some(c) => c,
                            None => break 'outer,
                        };
                        x += 1;
                        if c == '\n' {
                            x = 0;
                            y += 1;
                            tokens.get_mut().push(
                                Token { token: TokenType::COMMA, content: ",".to_string(), span: (y, x) });
                            break;
                        }
                    }
                    continue;
                } else {
                    tokens.get_mut().push(
                        Token { token: TokenType::NAME, content: name, span: (y, x) });
                }
                x += len;
                continue;
            }
            _ => {
                error(format!("lexer error {{ UnknownSym }} : y = {}, x = {}, c = {}", y, x, c), repl)
            }
        }
        x += 1;
    }
    tokens.get_mut().push(
        Token { token: TokenType::EOF, content: "".to_string(), span: (y, x) });
}

#[cfg(test)]
mod test_lexer {
    use std::cell::{Ref, RefCell};
    use crate::token::{Token, TokenType};

    #[test]
    fn scan_test() {
        let mut tokens: RefCell<Vec<Token>> = RefCell::new(Vec::new());
        crate::lexer::scan("::= hello    123.123 ".to_string(), &mut tokens, false);
        assert_eq!(RefCell::new(vec![Token { token: TokenType::ATOM, content: ":".to_string(), span: (0, 0) },
                                     Token { token: TokenType::BIND, content: ":=".to_string(), span: (0, 1) },
                                     Token { token: TokenType::NAME, content: "hello".to_string(), span: (0, 4) },
                                     Token { token: TokenType::NUM, content: "123.123".to_string(), span: (0, 14) },
                                     Token { token: TokenType::EOF, content: "".to_string(), span: (0, 15) }]),
                   tokens);
    }
}

File: depr/scala/src/main/scala/fungsi/src/Lexer.scala
package fungsi.src

import scala.annotation.tailrec

object Lexer:
  def scan(code: String): List[Token] =
    val symbols = "+-\\|/?.><!#@`^~%&*-_+="
    val keywords = Map("def" -> Token.DEF)

    @tailrec
    def HELP(tokens: List[Token], code: String): List[Token] =
      if code == "" then
        return (new Token(Token.EOF, "") +: tokens).reverse
      val c = code.head
      c match
        case '@' => HELP(new Token(Token.LAMBDA, "@") +: tokens, code.tail)
        case '$' => HELP(new Token(Token.QUOTE, "~") +: tokens, code.tail)
        case '^' => HELP(new Token(Token.RETURN, "^") +: tokens, code.tail)
        case '|' => HELP(new Token(Token.DELIMITER, "|") +: tokens, code.tail)
        case ';' => HELP(new Token(Token.SEMICOLON, ";") +: tokens, scipSpaces(code.tail))
        case ',' => HELP(new Token(Token.COMMA, ",") +: tokens, code.tail)
        case ':' =>
          if code.tail.head == '=' then
            HELP(new Token(Token.BIND, ":=") +: tokens, code.tail.tail)
          else
            val name = addName(code.tail, "")
            HELP(new Token(Token.ATOM, ":" + name._1) +: tokens, name._2)
        case '(' | '[' => HELP(new Token(Token.LPAREN, "(") +: tokens, code.tail)
        case ')' | ']' => HELP(new Token(Token.RPAREN, ")") +: tokens, code.tail)
        case '\r' | '\t' | '\n' | ' ' => HELP(tokens, code.tail)
        case _ if c.isDigit =>
          val num = addNum(code, "")
          HELP(new Token(Token.NUM, num._1) +: tokens, num._2)
        case _ if c.isLetter || symbols.contains(c) =>
          val name = addName(code, "")
          keywords get name._1 match {
            case Some(tok) if tok == Token.DEF =>
              HELP(new Token(keywords(name._1), name._1) +: tokens, name._2)
            case None | Some(_) =>
              HELP(new Token(Token.NAME, name._1) +: tokens, name._2)
          }
        case _ =>
          println("lexer error")
          Nil
    end HELP

    @tailrec
    def addSeq(code: String, res: String, end: Char): (String, String) =
      if code.head == end then
        return (res.reverse, code.tail)
      addSeq(code.tail, code.head +: res, end)
    end addSeq

    @tailrec
    def addNum(code: String, res: String): (String, String) =
      def isDigit(c: Char): Boolean = c.isDigit || c == '.'

      if !isDigit(code.head) then
        return (res.reverse, code)
      addNum(code.tail, code.head +: res)
    end addNum

    @tailrec
    def addName(code: String, res: String): (String, String) =
      def isLetter(c: Char): Boolean = c.isLetter || c.isDigit || symbols.contains(c)

      if !isLetter(code.head) then
        return (res.reverse.toLowerCase, code)
      addName(code.tail, code.head +: res)
    end addName

    @tailrec
    def scipSpaces(code: String): String =
      if code.head == '\n' then
        code.tail
      else
        scipSpaces(code.tail)

    HELP(Nil, code)
  end scan
end Lexer